src/simpleppl.jl

```diff
@@ -0,0 +1,213 @@
+using AbstractPPL
+import Base.getindex
+using SparseArrays
+using Setfield
+using Setfield: PropertyLens, get
+
+"""
+    ModelState(value::NamedTuple{T}, input::NamedTuple{T}, eval::NamedTuple{T}, kind::NamedTuple{T})
+
+Record the state of the model as a struct of NamedTuples, all
+sharing the same key values, namely, those of the model parameters.
+`value` should store the initial/current value of the parameters.
+`input` stores a tuple of inputs for a given node. `eval` are the
+anonymous functions associated with each node. These might typically
+be either deterministic values or some distribution, but could an 
+arbitrary julia program. `kind` is a tuple of symbols indicating
+whether the node is a logical or stochastic node.
+"""
+struct ModelState{T}
+    value::NamedTuple{T}
+    input::NamedTuple{T}
+    eval::NamedTuple{T}
+    kind::NamedTuple{T}
+end
+
+"""
+    DAG(inputs::NamedTuple{T})
+
+Struct containing the adjacency matrix for a particular model and 
+the topologically ordered vertex list.
+"""
+struct DAG
+    A::SparseMatrixCSC
+    sorted_vertex_list::Tuple
+    
+    function DAG(in::NamedTuple) 
+        inputs = keys(in)
+        A = adjacency_matrix(in) 
+        v_list = topological_sort_by_dfs(A)
+        new(A, inputs[v_list])
```

Comment:
```suggestion
    function DAG(inputs::NamedTuple) 
        input_names = keys(inputs)
        A = adjacency_matrix(inputs) 
        sorted_vertices = topological_sort_by_dfs(A)
        new(A, inputs[sorted_vertices])
```

Or something like that. At least, some better names.

Also, this should be an outer constructor. You can probably remove the type constraint, then it works also on dictionaries.

Then there should be some check of the inputs for consistency (existence of parent nodes etc.)

Reply:
And does the order of rows in `A` really correspond to the sorted order of vertices here? I think it better should, otherwise this looks a bit inconsistent.

Maybe you need to use [`SparseArrays.permute`](https://docs.julialang.org/en/v1/stdlib/SparseArrays/#SparseArrays.permute) after sorting.

Reply:
Thanks. The `A` matrix before did not correspond to the ordered vertex list -- nice catch. Fixed this now. Will work on adding some checks for correct graph structures.